### PR TITLE
odo catalog describe service <operator> --example

### PR DESCRIPTION
**Pull request by girishramnani**
**What type of PR is this?**
/kind feature
**What does this PR do / why we need it**:
see $SUBJECT

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/4796

**PR acceptance criteria**:

- [ ] Unit test 

- [X] Integration test 

- [ ] Documentation 

- [ ] Update changelog

- [X] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
Install an operator of your liking

```
odo catalog describe service postgresql-operator.v0.1.1/Database --example
apiVersion: postgresql.dev4devs.com/v1alpha1
kind: Database
metadata:
  name: database
spec:
  databaseCpu: 30m
  databaseCpuLimit: 60m
  databaseMemoryLimit: 512Mi
  databaseMemoryRequest: 128Mi
  databaseName: example
  databaseNameKeyEnvVar: POSTGRESQL_DATABASE
  databasePassword: postgres
  databasePasswordKeyEnvVar: POSTGRESQL_PASSWORD
  databaseStorageRequest: 1Gi
  databaseUser: postgres
  databaseUserKeyEnvVar: POSTGRESQL_USER
  image: centos/postgresql-96-centos7
  size: 1
```

### CODE REVIEW

**Comment by girishramnani** [comment]
Will update AC in a subsiquent PR

**Comment by girishramnani** [comment]
```
odo catalog describe service etcdoperator.v0.9.4-clusterwide/EtcdCluster --example
apiVersion: etcd.database.coreos.com/v1beta2
kind: EtcdCluster
metadata:
  annotations:
    etcd.database.coreos.com/scope: clusterwide
  name: example
spec:
  size: 3
  version: 3.2.13

 % odo catalog describe service etcdoperator.v0.9.4-clusterwide/EtcdCluster --example -o json
{
  "apiVersion": "etcd.database.coreos.com/v1beta2",
  "kind": "EtcdCluster",
  "metadata": {
    "annotations": {
      "etcd.database.coreos.com/scope": "clusterwide"
    },
    "name": "example"
  },
  "spec": {
    "size": 3,
    "version": "3.2.13"
  }
}
```

**Comment by dharmit** [comment]
>      % odo catalog describe service etcdoperator.v0.9.4-clusterwide/EtcdCluster --example -o json
>     {
>       "apiVersion": "etcd.database.coreos.com/v1beta2",
>       "kind": "EtcdCluster",
>       "metadata": {
>         "annotations": {
>           "etcd.database.coreos.com/scope": "clusterwide"
>         },
>         "name": "example"
>       },
>       "spec": {
>         "size": 3,
>         "version": "3.2.13"
>       }
>     }

@girishramnani I don't think this is the way we output JSON data for any command. Why are we deviating in this case?

**Comment by dharmit** [comment]
> Will update AC in a subsiquent PR - to avoid any unnecessary rebasing

Which acceptance criteria are you referring to?

**Review by dharmit — CHANGES_REQUESTED** [review]
We need to make changes to the tests to adhere to the flow proposed in #4770. Please change the tests accordingly for this PR.

**Comment by girishramnani** [comment]
> > Will update AC in a subsiquent PR - to avoid any unnecessary rebasing
> 
> Which acceptance criteria are you referring to?

sorry my mistake `wil update changelog`

**Comment by girishramnani** [comment]
> We need to make changes to the tests to adhere to the flow proposed in #4770. Please change the tests accordingly for this PR.

I feel its adhere to the flow, there is a before each already which runs the common part and I am just executing the "non-common" part of test in that `It` block.

**Comment by girishramnani** [comment]
> > ```
> >  % odo catalog describe service etcdoperator.v0.9.4-clusterwide/EtcdCluster --example -o json
> > {
> >   "apiVersion": "etcd.database.coreos.com/v1beta2",
> >   "kind": "EtcdCluster",
> >   "metadata": {
> >     "annotations": {
> >       "etcd.database.coreos.com/scope": "clusterwide"
> >     },
> >     "name": "example"
> >   },
> >   "spec": {
> >     "size": 3,
> >     "version": "3.2.13"
> >   }
> > }
> > ```
> 
> @girishramnani I don't think this is the way we output JSON data for any command. Why are we deviating in this case?

Could you elaborate?

**Comment by dharmit** [comment]
> > We need to make changes to the tests to adhere to the flow proposed in #4770. Please change the tests accordingly for this PR.
> 
> I feel its adhere to the flow, there is a before each already which runs the common part and I am just executing the "non-common" part of test in that `It` block.

And what about the rest of the test file? Is that file entirely adhering to that flow? If yes, can you update issue #4741 that this file doesn't need any changes to adhere to that flow?

The point is that it's not just about new tests being added to odo repo. We need to modify/refactor existing tests as well if we want the entire integration test suite to look like #4770.

**Comment by dharmit** [comment]
> Could you elaborate?

We're missing the GVKR info for JSON output.

If you do `odo catalog describe service <operator>/<cr> -o json`, you see ouptut like:
```json
{
        "kind": "CRDDescription",
        "apiVersion": "odo.dev/v1alpha1",
        "metadata": {
                "creationTimestamp": null
        },
        "spec": {}
}
```
What's proposed in this output is only JSON output of the ALM example. So it's more we're only showing the `spec` part.

**Comment by dharmit** [comment]
> > 
> 
> And what about the rest of the test file? Is that file entirely adhering to that flow? If yes, can you update issue #4741 that this file doesn't need any changes to adhere to that flow?
> 
> The point is that it's not just about new tests being added to odo repo. We need to modify/refactor existing tests as well if we want the entire integration test suite to look like #4770.

My bad on this one. I didn't notice that this was the same file as other OperatorHub tests. Sorry about the confusion.

File: pkg/odo/cli/catalog/describe/service.go [hunk 1]
```diff
@@ -39,7 +39,8 @@ type DescribeServiceOptions struct {
 	// generic context options common to all commands
 	*genericclioptions.Context
 
-	backend CatalogProviderBackend
+	backend   CatalogProviderBackend
+	IsExample bool
```
Comment (mik-dass) [inline pkg/odo/cli/catalog/describe/service.go:43]: This can be a non exported

Comment (girishramnani) [reply pkg/odo/cli/catalog/describe/service.go:43]: agreed, will change

**Comment by girishramnani** [comment]
this is how it looks like now 
```
{
  "kind": "OperatorExample",
  "apiVersion": "odo.dev/v1alpha1",
  "metadata": {
    "creationTimestamp": null
  },
  "spec": {
    "apiVersion": "etcd.database.coreos.com/v1beta2",
    "kind": "EtcdCluster",
    "metadata": {
      "annotations": {
        "etcd.database.coreos.com/scope": "clusterwide"
      },
      "name": "example"
    },
    "spec": {
      "size": 3,
      "version": "3.2.13"
    }
  }
}
```

**Comment by girishramnani** [comment]
human readable version looks like 
```
odo catalog describe service etcdoperator.v0.9.4-clusterwide/EtcdCluster --example        apiVersion: etcd.database.coreos.com/v1beta2
kind: EtcdCluster
metadata:
  annotations:
    etcd.database.coreos.com/scope: clusterwide
  name: example
spec:
  size: 3
  version: 3.2.13
```

File: pkg/odo/cli/catalog/describe/service.go (hunk 1 above)
Comment (girishramnani) [reply pkg/odo/cli/catalog/describe/service.go]: done, addressed the comment

**Comment by dharmit** [comment]
> this is how it looks like now
> 
> ```
> {
>   "kind": "OperatorExample",
>   "apiVersion": "odo.dev/v1alpha1",
>   "metadata": {
>     "creationTimestamp": null
>   },
>   "spec": {
>     "apiVersion": "etcd.database.coreos.com/v1beta2",
>     "kind": "EtcdCluster",
>     "metadata": {
>       "annotations": {
>         "etcd.database.coreos.com/scope": "clusterwide"
>       },
>       "name": "example"
>     },
>     "spec": {
>       "size": 3,
>       "version": "3.2.13"
>     }
>   }
> }
> ```

Thanks @girishramnani for this. When I looked at the output, it got me wondering if we should have a key in the `spec:` and put the alm example as its value. Something like:
```json
{
  "kind": "OperatorExample",
  "apiVersion": "odo.dev/v1alpha1",
  "metadata": {
    "creationTimestamp": null
  },
  "spec": {
    someKey: {
    "apiVersion": "etcd.database.coreos.com/v1beta2",
    "kind": "EtcdCluster",
    "metadata": {
      "annotations": {
        "etcd.database.coreos.com/scope": "clusterwide"
      },
      "name": "example"
    },
    "spec": {
      "size": 3,
      "version": "3.2.13"
    }
  }
}}

```

@kadel do you have any suggestion on this?

@girishramnani Sorry about being a pain on this PR.

File: pkg/odo/cli/catalog/describe/service.go (hunk 1 above)
Comment (girishramnani) [reply pkg/odo/cli/catalog/describe/service.go]: done

**Comment by girishramnani** [comment]
what should be the `somekey` though?

**Comment by girishramnani** [comment]
```
msg=Error: error creating Route53 Hosted Zone: TooManyHostedZones: Limits Exceeded: MAX_HOSTED_ZONES_BY_OWNER - Cannot create more hosted zones.
level=error msg=	status code: 400, request id: eabc56f1-58a7-4312-ab01-cd808a9fcaff
```

**Comment by girishramnani** [comment]
```
level=error
46
level=error msg=Error: [ERR]: Error building changeset: InvalidChangeBatch: [Tried to create resource record set [name='api.ci-op-kxqb4brl-89922.origin-ci-int-aws.dev.rhcloud.com.', type='A'] but it already exists]
```

**Comment by feloy** [comment]
/lgtm

As a user, it seems strange to me that the JSON content is not the same as the YAML one (but encapsulates the contents into an `OperatorExample` resource), but I understand we have to stick to the general JSON format

**Comment by kadel** [comment]
> Thanks @girishramnani for this. When I looked at the output, it got me wondering if we should have a key in the `spec:` and put the alm example as its value. Something like:

I think that the current format makes sense (no extra keys in `spec`)

**Comment by girishramnani** [comment]
Ack. Thank you.
Could you please take a look @dharmit

**Comment by dharmit** [comment]
> As a user, it seems strange to me that the JSON content is not the same as the YAML one (but encapsulates the contents into an `OperatorExample` resource), but I understand we have to stick to the general JSON format

Yeah. I agree on this one. But at the same time, I'm not sure if we should make an exception here because it's using the same flag (`-o json`) that's used elsewhere for JSON output.

/approve

**Pull request merged**